src/interfaces/Jsonapi.ts

```diff
@@ -0,0 +1,34 @@
+import { IModelConstructor, Model } from '@datx/core';
+import { IJsonapiModel } from '@datx/jsonapi';
+
+export type JsonapiResourceAttributes<TModel> = Partial<Omit<TModel, keyof IJsonapiModel | keyof Model>>;
+
+/**
+ * “Resource objects” appear in a JSON:API document to represent resources.
+ *
+ * @see https://jsonapi.org/format/#document-resource-objects
+ */
+export type JsonapiResource<TModelConstructor extends IModelConstructor> = {
+	id?: string;
+	lid?: string;
+	type: TModelConstructor['type'];
+	attributes: JsonapiResourceAttributes<InstanceType<TModelConstructor>>;
```

Comment:
This is not 100% correct.

```ts
class Model extends jsonapi(Model) {
  @Attribute({ toOne: () => ModelB })
  propA: ModelB;

  @Attribute({ map: "prop_b" })
  propB: string;
```

Both props are not going to be in a response but the typing will say that tey will exists. Do we really what this?

Reply:
ah crap, you are right. That's the limitation of having all the fields directly exposed on the model. We don't know which one is a relationship and which one is an attribute. :S 

🤔 is there a way to extract field information from the decorator? 
cc @DarkoKukovec 

Link to specification description of a field https://jsonapi.org/format/#document-resource-object-fields

Reply:
I have a hunch that it could be done with this plugin https://github.com/leonardfactory/babel-plugin-transform-typescript-metadata in combination with [Reflect.metadata](https://rbuckton.github.io/reflect-metadata/)

Reply:
When a property is a relationship, it will also have the referenced model id in `model.meta.refs[propertyName]` (and the property doesn't exist for attributes that are not references)

Reply:
Also, if a type is instanceof PureModel or an array of pure models, then it has to be a relationship.

Reply:
```tsx
export type JsonapiResourceAttributes<TModel> = Partial<
	Omit<
		{
			[Key in keyof TModel as TModel[Key] extends IJsonapiModel | Array<IJsonapiModel> ? never : Key]-?: TModel[Key];
		},
		keyof IJsonapiModel | keyof Model
	>
>;

export type JsonapiResourceRelationships<TModel> = Partial<
	Omit<
		{
			[Key in keyof TModel as TModel[Key] extends IJsonapiModel | Array<IJsonapiModel> ? Key : never]-?: TModel[Key];
		},
		keyof IJsonapiModel | keyof Model
	>
>;
```

This will handle the relationship part, but only that

Reply:
`JsonapiResourceRelationships` is not correct and I can't figure it out because I need a static class type of relationship but I have only instance type, for example, I need User.type literal form Session User relationship

Reply:
I'm also thinking to invest some time in CLI generator for extracting JSON:API model types to standalone file based on Model classes. 
@DarkoKukovec what do you think?